compose_rl/registry.py

```diff
@@ -1,43 +1,58 @@
 # Copyright 2024 MosaicML ComposeRL authors
 # SPDX-License-Identifier: Apache-2.0
 
-"""Registry for all the components in the Compose RL."""
+from llmfoundry.utils.registry_utils import create_registry
 
-from compose_rl.ppo.kl_controller import (
-    AdaptiveKLController,
-    BallKLController,
-    FixedKLController,
-    KLPIDController,
+_rewards_description = (
+    """The function rewards registry is used to register classes that implement function rewards.
+
+    One example of this is IncreasingNumbersReward. See reward_learning/functional.py for examples.
+
+    Args:
+        tokenizer: Tokenizer: The tokenizer being used.
+        kwargs: dict[Any, Any]: Additional keyword arguments.
+
+    Returns:
+        BaseReward: The reward class.
+    """
 )
-from compose_rl.reward_learning import (
-    BadGenerationEndReward,
-    ComposerHFPairwiseRewardModel,
-    ComposerMPTPairwiseRewardModel,
-    GSM8KAnswerVeriferReward,
-    GSM8KFormatVeriferReward,
-    IncreasingNumbersReward,
-    InferenceRewardModel,
-    MATHVerifierReward,
-    OutputLengthReward,
-    ShortResponseReward,
+rewards = create_registry(
+    'llmfoundry',
+    'rewards',
+    generic_type=None, # type: ignore
+    entry_points=True,
+    description=_rewards_description,
 )
 
-RL_REWARD_REGISTRY = {
-    'increasing_numbers': IncreasingNumbersReward,
-    'output_length': OutputLengthReward,
-    'short_response_reward': ShortResponseReward,
-    'inference_reward_model': InferenceRewardModel,
-    'mpt_pairwise': ComposerMPTPairwiseRewardModel,
-    'hf_pairwise': ComposerHFPairwiseRewardModel,
-    'bad_generation_end': BadGenerationEndReward,
-    'gsm8k_answer_verifier': GSM8KAnswerVeriferReward,
-    'gsm8k_format_verifier': GSM8KFormatVeriferReward,
-    'math_verifier': MATHVerifierReward,
-}
-
-KL_CONTROLLER_REGISTRY = {
-    'adaptive': AdaptiveKLController,
-    'fixed': FixedKLController,
-    'pid': KLPIDController,
-    'ball': BallKLController,
-}
+_kl_controller_description = (
+    """The KL Controller registry is used to register classes that implement KL Controller.
+
+    One example of this is FixedKLController. See ppo/kl_controller.py for examples.
+
+    Args:
+        kl_config: dict[Any, Any]: The config for the kl controller class.
+        device: Optional[torch.device]: The device to use for the kl controller.
+
+    Returns:
+        BaseKLController: The kl controller class.
+    """
+)
+kl_controllers = create_registry(
+    'llmfoundry',
+    'kl_controllers',
+    generic_type=None, # type: ignore
+    entry_points=True,
+    description=_kl_controller_description,
+)
+
+
+def initialize_registries():
+    """Initialize registries with their types after all modules are loaded."""
+    from compose_rl.ppo.kl_controller import BaseKLController
+    from compose_rl.reward_learning.base_reward import BaseReward
+
+    rewards.generic_type = type[BaseReward]  # type: ignore
```

Comment:
I don't think this really does anything cause the only point of the generic types are for static typing. The way I would do this is to create in `interfaces` folder in `compose_rl`, and move the `BaseKLController` and `BaseReward` there. The interfaces folder should not import from anywhere else in the library, to avoid circular imports.